1단계 - 엔티티 매핑

victor-jo opened the pull request:
1단계 - 엔티티 매핑 리뷰 요청드립니다!

잘 부탁드리겠습니다

victor-jo committed:
refactor(properties): DB 프로퍼티 파일 수정

- H2 DB 모드를 MySQL 로 수정
- format_sql 설정
- sql 보이도록 설정

victor-jo committed:
refactor(answer): JPA 어노테이션 적용

- SQL 스크립트와 일치하도록 적용

before:
    불일치
after:
    create table answer
    (
           id bigint not null auto_increment,
           contents longtext,
           created_at datetime(6) not null,
           deleted bit not null,
           question_id bigint,
           updated_at datetime(6),
           writer_id bigint,
           primary key (id)
    ) engine=InnoDB

victor-jo committed:
refactor(properties): H2 DB 에 사용되는 쿼리로 테스트 할 수 있도록 수정

- MySQL57Dialect 스크립트 모드 잠시 주석처리

victor-jo committed:
test(answer): 답변 엔티티 테스트

- save 및 1차 캐시 테스트

victor-jo committed:
refactor(deleteHistory): JPA 어노테이션 적용

- SQL 스크립트와 일치하도록 적용

before:
    불일치
after:
    create table delete_history
    (
        id            bigint not null auto_increment,
        content_id    bigint,
        content_type  varchar(255),
        create_date   datetime(6),
        deleted_by_id bigint,
        primary key (id)
    ) engine=InnoDB

victor-jo committed:
refactor(question): JPA 어노테이션 적용

- SQL 스크립트와 일치하도록 적용

before:
    불일치
after:
    create table question
    (
        id         bigint       not null auto_increment,
        contents   longtext,
        created_at datetime(6)  not null,
        deleted    bit          not null,
        title      varchar(100) not null,
        updated_at datetime(6),
        writer_id  bigint,
        primary key (id)
    ) engine=InnoDB

victor-jo committed:
refactor(user): JPA 어노테이션 적용

- SQL 스크립트와 일치하도록 적용

before:
    불일치
after:
    create table user
    (
        id         bigint      not null auto_increment,
        created_at datetime(6) not null,
        email      varchar(50),
        name       varchar(20) not null,
        password   varchar(20) not null,
        updated_at datetime(6),
        user_id    varchar(20) not null,
        primary key (id)
    ) engine=InnoDB

    alter table user
        add constraint UK_a3imlf41l37utmxiquukk8ajc unique (user_id)

victor-jo committed:
test(deleteHistory): 질문, 답변 삭제된 히스토리 테스트

victor-jo committed:
test(question): 질문 엔티티 테스트

victor-jo committed:
test(user): 사용자 엔티티 테스트

victor-jo committed:
refactor(app): 어플리케이션 실행되도록 H2Dialect 사용

minzzang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -3,13 +3,26 @@
     private Long id;
     private Long writerId;
     private Long questionId;
+    @Column(columnDefinition = "longtext")
```
clob 타입인데 longtext로 설정한 이유가 있을까요~?

victor-jo replied:
clob, blob 타입을 정의하는 @Lob 어노테이션 사용하였습니다

minzzang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -3,13 +3,26 @@
     private String contents;
+    @Column(nullable = false, columnDefinition = "datetime(6)")
+    private LocalDateTime createdAt = LocalDateTime.now();
+    @Column(columnDefinition = "datetime(6)")
```
datetime(6)으로 설정한 이유가 있나요~?

victor-jo replied:
주어진 MySQL 쿼리에 맞추느라 사용한 것 같습니다!

minzzang commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -32,6 +45,10 @@ public Answer(Long id, User writer, Question question, String contents) {
         this.contents = contents;
     }
 
+    protected Answer() {
```
접근 제어자 👍

minzzang commented on src/main/java/qna/domain/User.java:
```diff
@@ -2,18 +2,33 @@
+    @Id
+    @GeneratedValue(strategy = IDENTITY)
     private Long id;
+    @Column(unique = true, nullable = false, columnDefinition = "varchar(20)")
```
unique 설정 👍

minzzang commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,10 +1,26 @@
     private String contents;
     private Long writerId;
+    @Column(nullable = false, columnDefinition = "datetime(6)")
+    private LocalDateTime createdAt = LocalDateTime.now();
```
Auditing 기능을 사용해서 분리해보면 어떨까요~?

victor-jo replied:
Auditing 기능으로 부모 클래스로 분리하여 적용하였습니다!

minzzang commented on src/main/java/qna/domain/User.java:
```diff
@@ -2,18 +2,33 @@
     private String userId;
+    @Column(nullable = false, columnDefinition = "varchar(20)")
     private String password;
+    @Column(nullable = false, columnDefinition = "varchar(20)")
```
columnDefinition = "varchar(20)"과 length = 20과 어떤 차이가 있을까요?

victor-jo replied:
코드 속으로 들어가서 주석을 살펴보니 columnDefinition 은 DDL 생성에 맞추어지는 컬럼 타입이고
length 는 String 에서만 사용할 수 있는 컬럼 길이값 정의하는 거군요

감사합니다👍

minzzang commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,43 @@
 package qna.domain;
 
+import org.junit.jupiter.params.ParameterizedTest;
```
요구사항에` @DataJpaTest를 사용하여 학습 테스트를 해 본다` 라고 나와 있는데 CRUD를 학습할 수 있도록 save이외의 테스트 코드도 작성해보는건 어떨까요~?

victor-jo replied:
넵 예상되는 예외사항 및 CRUD 테스트 최대한 추가하여 보았습니다!

minzzang reviewed (changes requested):
안녕하세요 원준님~!

JPA 1단계 미션 잘 수행해주셨습니다! :+1:

소소한 코멘트와 질문 남겼는데 확인 부탁드립니다. :pray:

그럼 오늘 하루도 화이팅하세요!! :muscle:

victor-jo committed:
refactor(answer, question): Clob, Blob 타입에 대해 @Lob 타입으로 명시

before:
    @Column(columnDefinition = "longtext")
after:
    @Lob

victor-jo committed:
feat(jpaAuditingConfiguration): JPA Auditing 기능을 사용하기 위해 설정 클래스 추가

victor-jo committed:
feat(timeEntity): JPA Audit 기능 추가

- 생성일자, 수정일자 필드 중복제거 및 Audit 적용을 위한 엔티티 추가

victor-jo committed:
feat(all): JPA Audit 기능 적용

- 생성일자, 수정일자 필드 중복제거 및 Audit 적용

victor-jo committed:
test(answer): 답변 엔티티 테스트에 필요한 Fixture 함수명 변경

victor-jo committed:
refactor(properties): 로깅 레벨 조정

- SQL 쿼리에 바인딩 되는 값들을 보기 위한 레벨 조정

victor-jo committed:
test(answer): 답변 엔티티 내용 수정 테스트 케이스 추가

victor-jo committed:
refactor(timeEntity): 직접 생성해서 사용하지 않는 엔티티에 대해 추상클래스로 변경

victor-jo committed:
test(answer): 답변 엔티티 내용 삭제 테스트 케이스 추가

victor-jo committed:
test(fixture): 엔티티 테스트에 필요한 fixture 함수명 명시적으로 변경

victor-jo committed:
test(answer): 답변 엔티티 예외사항 테스트케이스 추가

victor-jo committed:
refactor(answerTest): 답변 엔티티 테스트 강제 flush 구문 삭제

victor-jo committed:
refactor(questionTest): 질문 엔티티 수정 테스트케이스 추가

- 질문이 등록되고 답변이 달리는 시나리오 테스트

victor-jo committed:
refactor(questionTest): 질문 엔티티 삭제 테스트케이스 추가

victor-jo committed:
refactor(answerTest): 같은 내용의 객체라도 아이디 값으로 객체를 구분하도록 수정

victor-jo committed:
refactor(question): 질문 엔티티에서 삭제된 질문에는 답변을 달 수 없도록 수정

victor-jo committed:
test(answerTest): 잘못 테스트되고 있는 케이스 제거

victor-jo committed:
refactor(user): 사용자 길이값 리팩터링

- 기존 columnDefinition 속성에서 length 로 변경

before:
    columnDefinition 으로 범용적으로 DB 타입 전체를 정의함
after:
    length 속성으로 문자열에서만 사용할 수 있는 컬럼길이 지정 속성을 사용함

victor-jo committed:
refactor(question): 사용자 길이값 리팩터링

victor-jo committed:
refactor(deleteHistory): 사용자 길이값 리팩터링

- 길이 기본값 '255' 로 불필요한 속성 제거

victor-jo commented:
피드백 감사합니다!
재리뷰 요청드립니다~

minzzang commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -1,22 +1,53 @@
+    @GeneratedValue(strategy = IDENTITY)
     private Long id;
+    @Enumerated(STRING)
+    @Column
```
속성을 정의하지 않는다면 없어도 될 것 같네요 ! 😄

minzzang commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,10 +1,22 @@
+    @Id
+    @GeneratedValue(strategy = IDENTITY)
     private Long id;
+    @Column(nullable = false, length = 100)
```
피드백 반영 👍

minzzang commented on src/main/java/qna/domain/TimeEntity.java:
```diff
@@ -0,0 +1,31 @@
+
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+public abstract class TimeEntity {
```
Auditing 👍

minzzang commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,83 @@
+    }
+
+    @Test
+    void 작성자가_없이_답변을_남길_수_없다() {
```
이 부분은 도메인 로직 테스트 같은데 domain과 repository 테스트를 분리하면 좋겠네요!

minzzang commented on src/test/java/qna/domain/UserTest.java:
```diff
@@ -1,6 +1,94 @@
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("rangeLimitArguments")
+    void 문자열_길이_검사(String representNamed, User newUser) {
```
꼼꼼한 테스트 👍

minzzang reviewed (approved):
안녕하세요 원준님~!

피드백 잘 반영해주셨습니다! :+1:

소소한 코멘트 남겼는데 확인 부탁드립니다. :pray:

그럼 다음 미션도 화이팅입니다!! :muscle:
